esbulk/esbulk.go

```diff
@@ -31,27 +32,56 @@ func (i *Indexer) Err() chan error { return i.err }
 // NewIndexer creates a new Elasticsearch bulk indexer. URL should be of the
 // form http://eshost:9200/_bulk.
 //
+// bufsz is the size of the upload buffer in kilobytes. bufsz < 1 will default
+// to 20mb.
+//
+// par is the number of parallel buffers to use. par < 1 will default to 3.
+//
 // Sends to docs should select on Indexer.Err to prevent deadlocking in case of
 // indexer error.
-func NewIndexer(url string, index string, docs <-chan *estypes.Doc) *Indexer {
+func NewIndexer(hosts []string, index string, bufsz, par int, docs <-chan *estypes.Doc) *Indexer {
 	indexer := &Indexer{
 		docs: docs,
 
-		// buffer of 1 to allow goroutine to exit before value is read
-		err: make(chan error, 1),
+		// buffer an error per parallel upload buffer
+		err: make(chan error, par),
+	}
+
+	if bufsz < 1 {
+		bufsz = 20 * 1024
+	}
+	if par < 1 {
+		par = 3
+	}
+	targets := make([]string, len(hosts))
+	for i, h := range hosts {
+		targets[i] = fmt.Sprintf("http://%s/_bulk", h)
 	}
+	ti := 0
 
 	go func() {
 		defer close(indexer.err)
 
```

Comment:
That sounds fine too. I think the only reason it ended up this way is because the code only grew a little bit at a time.

You reminded me to try everything with -race and sure enough I was using that `ti` variable unsafely!